src/shogun/lib/type_case.h

```diff
@@ -0,0 +1,369 @@
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/any.h>
+#include <shogun/lib/type_list.h>
+
+using namespace shogun;
+
+namespace shogun
+{
+	typedef Types<
+		bool, char, int8_t, uint8_t, int16_t, uint16_t, int32_t, uint32_t,
+		int64_t, uint64_t, float32_t, float64_t, floatmax_t, SGVector<int32_t>,
+		SGVector<int64_t>, SGVector<float32_t>, SGVector<float64_t>,
+		SGVector<floatmax_t>, SGMatrix<int32_t>, SGMatrix<int64_t>,
+		SGMatrix<float32_t>, SGMatrix<float64_t>, SGMatrix<floatmax_t>>::type
+		SG_TYPES;
+
+	enum class TYPE
+	{
+		PT_BOOL = 1,
+		PT_CHAR = 2,
+		PT_INT8 = 3,
+		PT_UINT8 = 4,
+		PT_INT16 = 5,
+		PT_UINT16 = 6,
+		PT_INT32 = 7,
+		PT_UINT32 = 8,
+		PT_INT64 = 9,
+		PT_UINT64 = 10,
+		PT_FLOAT32 = 11,
+		PT_FLOAT64 = 12,
+		PT_FLOATMAX = 13,
+		PT_SGOBJECT = 14,
+		PT_COMPLEX128 = 15,
+		PT_SGVECTOR_FLOAT32 = 16,
+		PT_SGVECTOR_FLOAT64 = 17,
+		PT_SGVECTOR_FLOATMAX = 18,
+		PT_SGVECTOR_INT32 = 19,
+		PT_SGVECTOR_INT64 = 20,
+		PT_SGMATRIX_FLOAT32 = 21,
+		PT_SGMATRIX_FLOAT64 = 22,
+		PT_SGMATRIX_FLOATMAX = 23,
+		PT_SGMATRIX_INT32 = 24,
+		PT_SGMATRIX_INT64 = 25,
+		PT_UNDEFINED = 26
+	};
+
+	enum class CONTAINER_TYPE
+	{
+		SCALAR = 1,
+		VECTOR = 2,
+		MATRIX = 3
+	};
+
+} // namespace shogun
+
+template <typename T>
+struct sg_primitive_type
+{
+};
+
+#define SG_PRIMITIVE_TYPE(T, ptype)                                            \
+	template <>                                                                \
+	struct sg_primitive_type<T>                                                \
+	{                                                                          \
+		static constexpr TYPE pvalue = ptype;                                  \
+	};
+
+SG_PRIMITIVE_TYPE(bool, TYPE::PT_BOOL)
+SG_PRIMITIVE_TYPE(char, TYPE::PT_CHAR)
+SG_PRIMITIVE_TYPE(int8_t, TYPE::PT_INT8)
+SG_PRIMITIVE_TYPE(uint8_t, TYPE::PT_UINT8)
+SG_PRIMITIVE_TYPE(int16_t, TYPE::PT_INT16)
+SG_PRIMITIVE_TYPE(uint16_t, TYPE::PT_UINT16)
+SG_PRIMITIVE_TYPE(int32_t, TYPE::PT_INT32)
+SG_PRIMITIVE_TYPE(uint32_t, TYPE::PT_UINT32)
+SG_PRIMITIVE_TYPE(int64_t, TYPE::PT_INT64)
+SG_PRIMITIVE_TYPE(uint64_t, TYPE::PT_UINT64)
+SG_PRIMITIVE_TYPE(float32_t, TYPE::PT_FLOAT32)
+SG_PRIMITIVE_TYPE(float64_t, TYPE::PT_FLOAT64)
+SG_PRIMITIVE_TYPE(floatmax_t, TYPE::PT_FLOATMAX)
+SG_PRIMITIVE_TYPE(complex128_t, TYPE::PT_COMPLEX128)
+SG_PRIMITIVE_TYPE(SGVector<float32_t>, TYPE::PT_SGVECTOR_FLOAT32)
+SG_PRIMITIVE_TYPE(SGVector<float64_t>, TYPE::PT_SGVECTOR_FLOAT64)
+SG_PRIMITIVE_TYPE(SGVector<floatmax_t>, TYPE::PT_SGVECTOR_FLOATMAX)
+SG_PRIMITIVE_TYPE(SGVector<int32_t>, TYPE::PT_SGVECTOR_INT32)
+SG_PRIMITIVE_TYPE(SGVector<int64_t>, TYPE::PT_SGVECTOR_INT64)
+SG_PRIMITIVE_TYPE(SGMatrix<float32_t>, TYPE::PT_SGMATRIX_FLOAT32)
+SG_PRIMITIVE_TYPE(SGMatrix<float64_t>, TYPE::PT_SGMATRIX_FLOAT64)
+SG_PRIMITIVE_TYPE(SGMatrix<floatmax_t>, TYPE::PT_SGMATRIX_FLOATMAX)
+SG_PRIMITIVE_TYPE(SGMatrix<int32_t>, TYPE::PT_SGMATRIX_INT32)
+SG_PRIMITIVE_TYPE(SGMatrix<int64_t>, TYPE::PT_SGMATRIX_INT64)
```

Comment:
this is not nice and quite long....
BUT the good thing is that we (shogun framework devs) do this once, and not people who write higher level framework code.

Reply:
I know.. The only other time that it might be used is if someone wants to create a custom map. Do you mean the list is long or the names?

Reply:
no just the big stack. But it is OK, as it is better than before